Use additional core constants for envisalink

RichieFrame opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Use additional core constants for envisalink, extends upon #46136

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RichieFrame committed:
Add more core constants

RichieFrame committed:
Add more core constants

RichieFrame committed:
Use more core constants

RichieFrame committed:
Use more core constants

RichieFrame committed:
Add vacation mode arm support for envisalink

Add vacation mode arm support, properly reference STATE_ALARM_ARMING for panels that return exit_delay prior to arm status

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev

RichieFrame committed:
Pending and arming states must be before armed modes

The delay flags exist simultaneously with the arm mode flags, so the delays must be detected first

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev

RichieFrame committed:
Add partition number to alarm keypad attributes

Also fix the default icon

RichieFrame committed:
Add partition number to alarm control panel attributes

RichieFrame committed:
Add partition configuration option for zones

RichieFrame committed:
Add partition configuration option for zones

Also reference configured zone number as an attribute in addition to partition number

RichieFrame committed:
black

RichieFrame committed:
sorting

RichieFrame committed:
isort

RichieFrame committed:
Faster zone close detection

When a zone closes, there can be a very large delay to when the Envisalink reports the closure. Adding the zone's partition number as an attribute, and using that to read the partition's keypad state allows detection of partition readyness indicating all zones are closed. If there are multiple open zones, the sensor reports closure of a single zone the same as before, but if it is the only open zone, the closure will be immediate

RichieFrame committed:
black, isort

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev

RichieFrame committed:
Add optional user list to integration

This gives the option of a configured list of user id codes and associated user names. When the keypad updates the last armed and disarmed user numbers, they will be translated to a user name if it exists.

Example yaml config entries for Honeywell:

  users:
    0:
      name: 'System'
    1:
      name: 'Installer'
    2:
      name: 'System Master'
    3:
      name: 'Partition 1 Master'

RichieFrame committed:
Add optional user list to integration

RichieFrame committed:
black, isort

RichieFrame committed:
Add code to partition configuration

Add codes to partitions
This allows either an override of the main configured code for a specific partition, or a partition to be configured with a code if there is no main configured code

Since configuring the main code allows arming (and potentially disarming) panels without code entry, configuring the code on a per partition basis requires code entry on other partitions, preserving desired security controls.

Different partitions may have different users, and different users may have different authority levels. A possible configuration is that a user from one partition may not arm or disarm other partitions. Another possibility is that a user may arm the system or partition, but not disarm, or that they may only disarm if their code was used to arm. Since there are numerous combinations, the integration should not be limited by a single code for all partitions.

Since user codes are logged and possibly reported to a central monitoring station, it may be undesirable to simply add an additional user with the correct authority to control the panels, since it is no longer clear who the actual person was who is performing security functions.

On the other hand, it may also be desirable to add a specific code for home automation, which is unable to disarm the system unless it was also used to arm it. That way HA users could easily arm and disarm the system without code entry, but arming with a different code at the keypad can lock that out.

RichieFrame committed:
Add code to partition configuration

RichieFrame committed:
Limit custom function registration to DSC configured panel type

Rename service schema and attributes to be more clear regarding DSC and inline with keypress service naming from alarm_control_panel.py, add check if panel is DSC before registering service

frenck commented on homeassistant/components/envisalink/__init__.py:
```diff
@@ -67,12 +77,13 @@
                 vol.Optional(CONF_PARTITIONS): {vol.Coerce(int): PARTITION_SCHEMA},
-                vol.Optional(CONF_EVL_PORT, default=DEFAULT_PORT): cv.port,
+                vol.Optional(CONF_USERS): {vol.Coerce(int): USERS_SCHEMA},
+                vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
```
We no longer allow integrations to add or change a platform YAML configuration.

- ADR-0010: <https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision>

See our developer documentation on how to get started creating a configuration flow for this integration:

<https://developers.home-assistant.io/docs/config_entries_config_flow_handler>

RichieFrame committed:
Add code for activating output devices on Honeywell

Function is similar to command output on DSC panels

The system may have multiple users with limited permissions, and their code may not have the authority to activate output devices. However, they may still need access to this function without giving them the system master code. Adding an optional code for activating devices keeps the code secret while still allowing device activation.

An example is using a panel connected relay to open and close garage doors, turn on lights, or mute a doorbell.

By configuring an additional code and adding an additional function to control output devices, their activation can be controlled by scripts and automations without needing to expose the system master code.

Additionally, sending the command to a keypad on a different partition may be desirable, as all keypads on that partition will chime during device activation, even if chime mode is off.

RichieFrame committed:
Add code for activating output devices on Honeywell

RichieFrame committed:
Adds attributes for the state of DPMS and screensaver

RichieFrame committed:
typos, isort, black

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev

RichieFrame committed:
remove trailing spaces

RichieFrame committed:
formatting, more consistent function code

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev

RichieFrame committed:
Merge branch 'home-assistant:dev' into dev
